### PR TITLE
Implement arguments exotic objects

### PR DESCRIPTION
**Pull request by jedel1043**
This Pull Request builds from #1354 and #1518 to implement arguments exotic objects.

It changes the following:

- Implements argument exotic objects
- Refactors `ClosureFunction` to hold external context as a `NativeObject`
- Adds a new `FunctionBuilder::closure_with_captures` function

Should I extract the last two changes as their own PR?.

Drafting this because right now we don't have a way to know if we are `call_construct`ing in strict mode, so a lot of tests fail because they expect an unmapped argument ordinary object in strict mode and a mapped argument exotic object in non-strict mode.

### CODE REVIEW

**Comment by jedel1043** [comment]
> > Should I extract the last two changes as their own PR?.
> 
> I would be happy to have smaller change-sets to review xD

Okay! I'm still gonna draft it until #1518 merges just in case

**Comment by jedel1043** [comment]
Now that #1550 merged, this is just waiting for another review on #1523 in order to merge it.

File: boa/src/object/gcobject.rs
Comment (jedel1043) [inline boa/src/object/gcobject.rs:242]: @raskad Do we need to check both `context` and `body` for strictness? If we do, can we pass the global strictness into all functions? Because that would remove the context check

**Comment by jedel1043** [comment]
### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 80,918 | 80,918 | 0 |
  | Passed | 33,178 | 33,280 | **+102** |
  | Ignored | 15,898 | 15,898 | 0 |
  | Failed | 31,842 | 31,740 | **-102** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 41.00% | 41.13% | **+0.13%** |
  
  <details><summary><b>Fixed tests (102):</b></summary>
  
  ```
  test/language/statements/for-of/arguments-mapped-aliasing.js (previously Failed)
  test/language/statements/for-of/arguments-mapped.js (previously Failed)
  test/language/arguments-object/10.6-11-b-1.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-11-b-1.js (previously Failed)
  test/language/arguments-object/10.6-12-2.js (previously Failed)
  test/language/arguments-object/10.6-7-1.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-7-1.js (previously Failed)
  test/language/arguments-object/10.6-13-a-3.js (previously Failed)
  test/language/arguments-object/10.6-13-c-1-s.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-5-1.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-5-1.js (previously Failed)
  test/language/arguments-object/10.6-14-c-4-s.js [strict mode] (previously Failed)
  test/language/arguments-object/S10.6_A5_T1.js [strict mode] (previously Failed)
  test/language/arguments-object/S10.6_A5_T1.js (previously Failed)
  test/language/arguments-object/S10.6_A2.js [strict mode] (previously Failed)
  test/language/arguments-object/S10.6_A2.js (previously Failed)
  test/language/arguments-object/S10.6_A3_T1.js [strict mode] (previously Failed)
  test/language/arguments-object/S10.6_A3_T1.js (previously Failed)
  test/language/arguments-object/10.6-13-c-2-s.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-13-c-2-s.js (previously Failed)
  test/language/arguments-object/10.6-13-a-1.js (previously Failed)
  test/language/arguments-object/10.6-14-c-1-s.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-14-c-1-s.js (previously Failed)
  test/language/arguments-object/10.6-10-c-ii-1.js (previously Failed)
  test/language/arguments-object/S10.6_A4.js (previously Failed)
  test/language/arguments-object/10.6-13-a-2.js (previously Failed)
  test/language/arguments-object/10.6-10-c-ii-2.js (previously Failed)
  test/language/arguments-object/10.6-13-c-3-s.js [strict mode] (previously Failed)
  test/language/arguments-object/10.6-2gs.js [strict mode] (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-delete-3.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-nonwritable-5.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-nonwritable-3.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-delete-2.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-2.js (previously Failed)
  test/language/arguments-object/mapped/nonconfigurable-nonenumerable-nonwritable-descriptors-set-by-param.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-3.js (previously Failed)
  test/language/arguments-object/mapped/nonconfigurable-nonenumerable-nonwritable-descriptors-set-by-arguments.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-delete-4.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-4.js (previously Failed)
  test/language/arguments-object/mapped/writable-enumerable-configurable-descriptor.js (previously Failed)
  test/language/arguments-object/mapped/mapped-arguments-nonconfigurable-nonwritable-4.js (previously Failed)
  test/built-ins/String/prototype/trim/15.5.4.20-2-51.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/trim/15.5.4.20-2-51.js (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-15.js [strict mode] (previously Failed)
  test/built-ins/Object/create/15.2.3.5-4-15.js (previously Failed)
  test/built-ins/Object/keys/15.2.3.14-6-4.js [strict mode] (previously Failed)
  test/built-ins/Object/keys/15.2.3.14-6-4.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-293-2.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-296-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-339-4.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-339-4.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-293-3.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-43-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-43-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-540-10.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-540-10.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-531-7.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-531-7.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-258-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-258-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-149-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-149-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-175-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-175-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-228-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-228-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-292-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-339-2.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-339-2.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-531-16.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-531-16.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-96-1.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-3-96-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-294-1.js (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-540-5.js [strict mode] (previously Failed)
  test/built-ins/Object/defineProperty/15.2.3.6-4-540-5.js (previously Failed)
  test/built-ins/Object/prototype/toString/Object.prototype.toString.call-arguments.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/toString/Object.prototype.toString.call-arguments.js (previously Failed)
  test/built-ins/Object/getPrototypeOf/15.2.3.2-2-28.js [strict mode] (previously Failed)
  test/built-ins/Object/getPrototypeOf/15.2.3.2-2-28.js (previously Failed)
  test/built-ins/Object/preventExtensions/15.2.3.10-3-11.js [strict mode] (previously Failed)
  test/built-ins/Object/preventExtensions/15.2.3.10-3-11.js (previously Failed)
  test/built-ins/Object/preventExtensions/15.2.3.10-3-21.js [strict mode] (previously Failed)
  test/built-ins/Object/preventExtensions/15.2.3.10-3-21.js (previously Failed)
  test/built-ins/ArrayIteratorPrototype/next/args-mapped-expansion-before-exhaustion.js (previously Failed)
  test/built-ins/ArrayIteratorPrototype/next/args-mapped-expansion-after-exhaustion.js (previously Failed)
  test/built-ins/ArrayIteratorPrototype/next/args-mapped-truncation-before-exhaustion.js (previously Failed)
  test/built-ins/ArrayIteratorPrototype/next/args-mapped-iteration.js (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/filter/15.4.4.20-1-15.js (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduceRight/15.4.4.22-1-15.js (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/reduce/15.4.4.21-1-15.js (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/map/15.4.4.19-1-15.js (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/some/15.4.4.17-1-15.js (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/every/15.4.4.16-1-15.js (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-1-15.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/forEach/15.4.4.18-1-15.js (previously Failed)
  ```

File: boa/src/object/gcobject.rs
Comment (raskad) [reply boa/src/object/gcobject.rs:242]: I think in theory we could pass the global strictness in all functions. But my proposal would be to wait with the strictness refactor until we have some more errors that depend on strictness implemented. Then we can be somewhat sure that we do not break anything and we know how the apis to manage strictness should look.

File: boa/src/object/internal_methods/arguments.rs
Comment (raskad) [inline boa/src/object/internal_methods/arguments.rs:219]: The spec does not expect `set` to fail here. So we should either `expect` on `set` or assert that `set_status` is `Ok(true)`.

File: boa/src/object/internal_methods/arguments.rs
Comment (raskad) [inline boa/src/object/internal_methods/arguments.rs:134]: The spec does not expect `set` to fail here. So we should either `expect` on `set` or assert that `set_status` is `Ok(true)`.

**Review by raskad — CHANGES_REQUESTED** [review]

File: boa/src/builtins/function/arguments.rs
Comment (jedel1043) [reply boa/src/builtins/function/arguments.rs:147]: Missed the optimization, thanks!

File: boa/src/object/gcobject.rs
Comment (jedel1043) [reply boa/src/object/gcobject.rs:213]: Noted

**Review by raskad — APPROVED** [review]

**Review by RageKnify — APPROVED** [review]

**Review by Razican — APPROVED** [review]

**Pull request merged**
